File: routines/source.ts
// Source stage executor (skeleton)
// Mirrors compose_with_options() rules to produce ordered -f files and resolved services/options.

export type SourceOutput = { files: string[]; services: string[]; options: string[] };

export async function executeSource(_profile = "default"): Promise<SourceOutput> {
  // TODO: detect capabilities (nvidia, cdi, rocm, mdc)
  // TODO: include base compose.yml and overlays per rules
  return { files: ["compose.yml"], services: [], options: ["*"] };
}

File: core/source.py
from __future__ import annotations
from dataclasses import dataclass
from typing import List
from .descriptor import Descriptor, Profile

@dataclass
class SourceOutput:
    files: List[str]
    services: List[str]
    options: List[str]

def execute(profile_name: str, d: Descriptor) -> SourceOutput:
    # TODO: Detect capabilities (nvidia, cdi, rocm, mdc)
    prof: Profile = d.profiles.get(profile_name, Profile())
    files: List[str] = ["compose.yml"]
    # TODO: Include overlays per rules (service, option, cross-files)
    return SourceOutput(files=files, services=prof.services, options=prof.options)
